Throttled and unthrottled editing

shresthabijay opened the pull request:
refers to #330

shresthabijay commented:
I have separated logic from previous `onChangeHandler` into two different functions. `thoughtChangeHandler` and `onChangeHandler`. Now `thoughtChangeHandler` is only responsible for dispatch calls and tutorial logic. And `onChangeHandler` performs Readonly attributes checks and type validation . And with the update it now checks for the change of superscript and availability of URL to decide calling `thoughtChangeHandler` immediately or using throttled function. Please review the changes and let me know what you think.

raineorshine commented:
This isn't quite working for me yet. 

If I edit a thought that has a superscript, the superscript is sometimes updated immediately and sometimes not.

Also when I have the following structure:

- Boolean
  - `=options`
    - true
    - false
  - true

When I edit `A.true`, it fails to update the validation error immediately, such as when I change `true` to `tru` and back to `true` quickly. If I select the text and type "false" over it quickly, the validation message says `Invalid Value: "f" and does not change.

shresthabijay commented:
> When I edit `A.true`, it fails to update the validation error immediately, such as when I change `true` to `tru` and back to `true` quickly. If I select the text and type "false" over it quickly, the validation message says `Invalid Value: "f" and does not change.

When a valid option is passed to `thoughtChangeHandler` we are not setting error to `null` . That is the reason why validation error persists even after we provide valid option. 

Also you can notice that for a meta `=options` when we provide a valid option and start editing it, the superscript doesn't change. It is because validation is happening on the scope of `Editable` and `existingThoughtChange` is not called for invalid option. `Superscript` component only knows the change in context length when `existingThoughtChange` is called.

shresthabijay commented:
I have updated the logic that was causing inconsistent immediate updates . Also I set error to null for inside `thoughtChangeHandler` to update the validation error after valid option has been passed. Please review the changes !

shresthabijay commented:
> Can you also make the bullet and text turn red when it has an invalid value using the [`invalid-option`](https://github.com/cybersemics/em/blob/a20b71b9cb7f8d3551240b205a849445f5ad09a5/src/components/Thought.js#L276) className?
> 
This is a tricky situation. I created a local state `invalidState`  to pass `invalid-option` class to the `ContentEditable` component. It works properly setting text color to red for invalid option but the problem is that current implementation depends on `Editable` to avoid re-rendering on change. But using local state causes re-rendering eventually causing input to loose focus and on back key press while editing often causes two thoughts to merge. 

We need to think of a proper way of using local state with this component. Biggest problem is loosing focus on re-render. I will get back to you with my ideas to address this issue. Please let me know if you have any solution.

raineorshine commented:
Yup, exactly. The `Editable` cannot re-render while it is being edited. Class names can change dynamically though without the whole element re-rendering I believe.

shresthabijay commented:
Hi Raine. I I have made following changes.

1. I used a `ref` to access `ContentEditable` node and used it add and remove `invalid-option` class to it. 

2. I used another `ref` called `annotationRef` on `Thought` component and passed it to `AnnotationThought` as `innerRef` prop to access it's node.

3. I created a `memoized` callback function called `setSuperscriptVisibilty` that sets visibility of `AnnotationThought` **(superscript)** using  `annotationRef` to change it's opacity. This function is passed to `Editable` to control visibility of **superscript** when invalid option is passed. It is important to make it a `memoized` callback in order to maintain **referential equality** to avoid re-rendering of `Editable` component. 

4. I also set `innerHTML` of `ContentEditable` to latest valid option , removed validation error and `invalid-option` class on input  blur.

Please review the changes. Thank you!

raineorshine commented:
Thank you for your time and work on this. I appreciate the creative solution!

After giving this some thought, I think we have to look for another solution. I'm just not comfortable resorting to so many refs and DOM manipulation. It might solve this problem, but it is bound to cause issues in the future.

I think we need to stick with the one-way data flow of Redux. There should be a real-time stream of `thoughtChange` events and a throttled stream of `thoughtFlush` events that defer the heavier operations such as updating `localForage`, syncing with `Firebase`, and updating `recentlyEdited`.

I value your opinion and look forward to hearing your thoughts. Thanks!

shresthabijay commented:
> After giving this some thought, I think we have to look for another solution. I'm just not comfortable resorting to so many refs and DOM manipulation. It might solve this problem, but it is bound to cause issues in the future.
> 

I agree! But with current architecture there are many situation where we have to persist data for full lifetime of the component (between component re-renders) . For these cases using `useRef` is by far the best solution provided by React. Also I had to use `DOM` manipulation because using state would cause component re-render causing input to loose focus. 

With current architecture understanding life-cycle of components like `SubThought` , `Thoughts` is quite complex . Specially it is hard to predict cause of re-renders and also how re-rendering affects the overall application. With more features it would be even harder to maintain the component. I strongly suggest we find a way to make component life cycle more predictable on change of application state. 

> I think we need to stick with the one-way data flow of Redux. There should be a real-time stream of `thoughtChange` events and a throttled stream of `thoughtFlush` events that defer the heavier operations such as updating `localForage`, syncing with `Firebase`, and updating `recentlyEdited`.
> 
Yes I agree ! We need a way to throttle redux actions instead of throttling event handlers that calls dispatch functions. I was thinking of writing a `middleware` that can throttle incoming actions if needed. I did some research and found this library https://github.com/mathieudutour/redux-throttle#readme . I think this will be a good approach. I am not sure to use this library or create custom middleware.  I will test this library and get back to you. Let me know what you think!

raineorshine commented:
> > After giving this some thought, I think we have to look for another solution. I'm just not comfortable resorting to so many refs and DOM manipulation. It might solve this problem, but it is bound to cause issues in the future.
> 
> I agree! But with current architecture there are many situation where we have to persist data for full lifetime of the component (between component re-renders) . For these cases using `useRef` is by far the best solution provided by React. 

Yes, which I am okay doing internally within a single component. However passing refs in props to other components seems dangerous.

> Also I had to use `DOM` manipulation because using state would cause component re-render causing input to loose focus.

Only if the component is subscribed to the changed state through `connect` or `useSelector`. I have been very careful to ensure that the `Editable` component does not re-render when its thought's value is changed.

> With current architecture understanding life-cycle of components like `SubThought` , `Thoughts` is quite complex . Specially it is hard to predict cause of re-renders and also how re-rendering affects the overall application. With more features it would be even harder to maintain the component. I strongly suggest we find a way to make component life cycle more predictable on change of application state.

Yes, I agree

> > I think we need to stick with the one-way data flow of Redux. There should be a real-time stream of `thoughtChange` events and a throttled stream of `thoughtFlush` events that defer the heavier operations such as updating `localForage`, syncing with `Firebase`, and updating `recentlyEdited`.
> 
> Yes I agree ! We need a way to throttle redux actions instead of throttling event handlers that calls dispatch functions. I was thinking of writing a `middleware` that can throttle incoming actions if needed. I did some research and found this library https://github.com/mathieudutour/redux-throttle#readme . I think this will be a good approach. I am not sure to use this library or create custom middleware. I will test this library and get back to you. Let me know what you think!

Yes, this doesn't look bad. I like the idea of dispatching a FLUSH action instead of having to maintain the throttle reference.

raineorshine commented:
If there were a separate state variable `editingValue` and a corresponding action `setEditingValue`, then the `Superscript` and `ThoughtAnnotation` components could add `editingValue` to its `mapStateToProps`, and update without the full `existingThoughtChange` action. It just requires some logic to replace the normal `value` with `editingValue`. This avoids DOM manipulation and uses a predictable data pathway.

shresthabijay commented:
> If there were a separate state variable `editingValue` and a corresponding action `setEditingValue`, then the `Superscript` and `ThoughtAnnotation` components could add `editingValue` to its `mapStateToProps`, and update without the full `existingThoughtChange` action. It just requires some logic to replace the normal `value` with `editingValue`. This avoids DOM manipulation and uses a predictable data pathway.

It's a good idea given that we want to avoid DOM manipulation inside `ThoughtAnnotation`. But if  we provide only editingValue`  to the components we will need logic to know if we have to hide superscript inside `ThoughtAnnotation`  (=meta validations).  For example we need `options` data for meta(=options) validation. That logic is already inside `Editable`. Also we need to make superscript visible on events like `onBlur` of `ContentEditable`. What do you think ?

shresthabijay commented:
> > Also I had to use `DOM` manipulation because using state would cause component re-render causing input to loose focus.
> 
> Only if the component is subscribed to the changed state through `connect` or `useSelector`. I have been very careful to ensure that the `Editable` component does not re-render when its thought's value is changed.

We need to toggle `invalid-option` class to `ContentEditable` .  `useState` hook cannot be used because it will cause `Editable` to re-render. So without any internal state of component , we only have option to manipulate `DOM` to toggle `invalid-option` class.

raineorshine commented:
> if we provide only `editingValue` to the components we will need logic to know if we have to hide superscript inside `ThoughtAnnotation`

Hiding the superscript was a subpar workaround. If we `connect` the `Superscript` component to `editingValue`, then it can update its `numContexts` in real-time.

> We need to toggle `invalid-option` class to `ContentEditable` .  `useState` hook cannot be used because it will cause `Editable` to re-render. So without any internal state of component , we only have option to manipulate `DOM` to toggle `invalid-option` class.

Yes, that's right. As long as the ref is contained within the component that is a fine way to make DOM updates without re-rendering the component.

The bullet should also receive `invalid-option`. It should be able to `connect` to `editingValue` for this.

shresthabijay commented:
Hi Raine! I have added `editingValue`  state in `redux` state to get real time editing value of currently edited thought. I also added `invalidState` to get real time meta validation error which is updated by `Editable` handlers. `editingValue` is updated by `onChangeHandler` inside `Editable`. `invalidState` is updated to true if any meta validation occurs inside `onChangeHandler`. 

`editingValue` alone can only help in obtaining real time context length . But with `invalidState` real time meta validation error status we don't have to write another logic for determining if `editingValue` has any validation error. We can easily set `invalid-option` class to `Bullet` component using `invalidState`. Also I have used both `invalidState` and `editingValue`  to update superscript of a thought in real-time only if thought has any validation error. 

It is working properly. Only thing not working is the offset of superscript is too much when context length is being updated in real time with (=options) meta validation error. I can't figure out how to control that offset. 

Please review it and let me know what you think of this approach.

shresthabijay commented on src/components/ThoughtAnnotation.js:
```diff
@@ -43,16 +43,21 @@ export const ThoughtAnnotation = connect(({ cursor, cursorBeforeEdit, focusOffse
-    focusOffset
+    focusOffset,
+    editingValue,
+    invalidState
```
Agreed!

shresthabijay commented on src/components/ThoughtAnnotation.js:
```diff
@@ -80,8 +85,9 @@ export const ThoughtAnnotation = connect(({ cursor, cursorBeforeEdit, focusOffse
-            minContexts === 0 || subthought.contexts.length > (subthought.text === value ? 1 : 0)
-              ? <StaticSuperscript n={subthought.contexts.length} />
+            // with real time context update we increase context length by 1
+            minContexts === 0 || subthought.contexts.length + (isRealTimeContextUpdate ? 1 : 0) > (subthought.text === value ? 1 : 0)
```
Sure !

shresthabijay commented on src/components/ThoughtAnnotation.js:
```diff
@@ -43,16 +43,21 @@ export const ThoughtAnnotation = connect(({ cursor, cursorBeforeEdit, focusOffse
   // disable intrathought linking until add, edit, delete, and expansion can be implemented
   // get all subthoughts and the subthought under the selection
+
+  const isRealTimeContextUpdate = isEditing && invalidState && editingValue !== null // only show real time update if being edited while having meta validation error
```
> I'm not sure I understand the reason for `&& invalidState`. Don't we want to update the `contexts` regardless of whether it is in a valid or invalid state?

When thought is in a `invalid-state` we know that `exitingThoughtChange` is not called. So the `numContext` of the `editingValue` should be increased by 1. But in a valid state `exitingThoughtChange` is called with  `newValue` and `numContext` is already in sync for currently edited thought. If we don't use `invalidState` flag , superscript will always be increased by 1 for edited thought.

raineorshine replied:
Okay, that makes sense. Maybe add a comment like: 

`// do not increase numContexts when in an invalid state since the thought has not been updated in state`

shresthabijay commented on src/components/Editable.js:
```diff
@@ -188,15 +166,65 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
+
+    // run the thoughtChangeHandler immediately if superscript changes or it's a url (also when it changes true to false)
+    if (contextLengthChange || urlChange) {
+      // updating new supercript value and url boolean
```
Sure!

shresthabijay commented on src/components/Editable.js:
```diff
@@ -188,15 +166,65 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
+    // run the thoughtChangeHandler immediately if superscript changes or it's a url (also when it changes true to false)
+    if (contextLengthChange || urlChange) {
+      // updating new supercript value and url boolean
+      throttledChangeRef.current.flush() // flusing any possible updates
```
Agreed!

shresthabijay commented on src/components/Editable.js:
```diff
@@ -275,9 +307,16 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
+      setInvalidState(false)
+      error(null)
+
+      // When user provides invalid option (meta =options) and clicks outside , we reset the innerHTML to previous valid value
```
got it!

shresthabijay commented on src/action-creators/setEditingValue.js:
```diff
@@ -0,0 +1,4 @@
+import { store } from '../store'
+
+// real time editing value
+export const setEditingValue = value => store.dispatch({ type: 'editingValue', value })
```
Yes we should avoid unnecessary actions.

shresthabijay commented on src/components/Editable.js:
```diff
@@ -188,15 +166,65 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
+    const newNumContext = getContexts(newValue).length
+    const isNewValueURL = isURL(newValue)
+    const contextLengthChange = newNumContext > 0 || newNumContext !== getContexts(oldValueRef.current).length
+    const urlChange = isNewValueURL || isNewValueURL !== isURL(oldValueRef.current)
```
> I was able to remove two refs by deriving the URL and numContexts from `oldValueRef` in [12d0f6f](https://github.com/cybersemics/em/commit/12d0f6f045bf9c5135dbc34186f2f4c56bda6271). Seems cleaner to me. What do you think?

That's great! It better and much cleaner.

shresthabijay replied:
Hey Raine! I just found one minor mistake in the logic. With the change `thoughtChangeHandler` is being called immediately. To fix this we should deccrease the `oldValue` context length by 1.

```
const contextLengthChange = newNumContext > 0 || newNumContext !== getContexts(oldValueRef.current).length - 1
```

It is because after `existingThoughtChange` context length for the `oldValue` will be increased by 1.

shresthabijay commented on src/components/Editable.js:
```diff
@@ -188,15 +166,65 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
     throttledChangeRef.current.flush()
   })
 
-  useEffect(() => throttledChangeRef.current.flush, []) // clean up function when component unmounts (flushing throttle change)
```
Editable doesn't re-render and doesn't `unmount` in its life cycle. Only time it unmounts is when it's parent thought is not rendered.  Flush on `onBlur` event is enough. Still we can keep the unmount flush. Should we do that ?

raineorshine replied:
Yes, let's keep it to be safe. I'm not sure `onBlur` will necessarily be called every time this component unmounts, and missing a flush would be disastrous.

shresthabijay replied:
> @shresthabijay I'm getting a very strange error when I try to run `npm start`:
> 
> ```
> ./src/reducers/editingValue.js
>   Line 0:  Parsing error: ENOENT: no such file or directory, stat '/users/raine/projects/em/tsconfig.eslint.json'
> ```
> 
> It runs fine on the `dev` branch, and this PR is up-to-date with `dev`. The error itself is nonsensical since `/tsconfig.eslint.json` does exist. Not to mention that it doesn't seem to have anything to do with `./src/reducers/editingValue.js`.
> 
> Would be so grateful for you to investigate!

Hi Raine! It's a very strange issue. This PR is up-to-date with dev and has got all the changes related to typescript configuration. It works fine for me without any error. I tried looking for any possible ts configuration error but there are none. Have you tried updating the `typescript ` package to the latest version `^3.7.5` ?

shresthabijay commented on src/components/Editable.js:
```diff
@@ -275,9 +303,16 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
+      // on blur removing error, removing invalid-option class and making superscript visible
+      setContentInvalidState(false)
+      setInvalidState(false)
+      error(null)
```
Sure !

shresthabijay commented on src/components/Editable.js:
```diff
@@ -188,15 +166,65 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
   })
 
-  useEffect(() => throttledChangeRef.current.flush, []) // clean up function when component unmounts (flushing throttle change)
+  const onChangeHandler = e => {
```
Yes agreed!

shresthabijay commented on src/components/Editable.js:
```diff
@@ -275,9 +307,16 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
       }
     }}
     onBlur={() => {
+      // on blur removing error, removing invalid-option class and making superscript visible
```
Sure

raineorshine commented:
@shresthabijay I'm getting a very strange error when I try to run `npm start`:

```
./src/reducers/editingValue.js
  Line 0:  Parsing error: ENOENT: no such file or directory, stat '/users/raine/projects/em/tsconfig.eslint.json'
```

It runs fine on the `dev` branch, and this PR is up-to-date with `dev`. The error itself is nonsensical since `/tsconfig.eslint.json` does exist. Not to mention that it doesn't seem to have anything to do with `./src/reducers/editingValue.js`.

Would be so grateful for you to investigate!

raineorshine commented:
@anmolarora1 Maybe you have some input on the [above](https://github.com/cybersemics/em/pull/386#issuecomment-601876798) since it seems to involve `tsconfig.eslint.json`. I thought maybe 8ec877ae9abe21fe7319e71f6be3d1039ed2dfb9 caused it, but I changed reverted it and it didn't help (not to mention `dev` is working fine.

shresthabijay commented:
> @shresthabijay I'm getting a very strange error when I try to run `npm start`:
> 
> ```
> ./src/reducers/editingValue.js
>   Line 0:  Parsing error: ENOENT: no such file or directory, stat '/users/raine/projects/em/tsconfig.eslint.json'
> ```
> 
> It runs fine on the `dev` branch, and this PR is up-to-date with `dev`. The error itself is nonsensical since `/tsconfig.eslint.json` does exist. Not to mention that it doesn't seem to have anything to do with `./src/reducers/editingValue.js`.
> 
> Would be so grateful for you to investigate!

Hi Raine! It's a very strange issue. This PR is up-to-date with dev and has got all the changes related to typescript configuration. It works fine for me without any error. I tried looking for any possible ts configuration error but there are none. Have you tried updating the `typescript ` package to the latest version `^3.7.5` ?

raineorshine commented:
@shresthabijay Updated my `npm`, deleted `npm_modules` and `package-lock.json`, merged with `dev`, and it is working now!

raineorshine commented:
> > @shresthabijay I'm getting a very strange error when I try to run `npm start`:
> > ```
> > ./src/reducers/editingValue.js
> >   Line 0:  Parsing error: ENOENT: no such file or directory, stat '/users/raine/projects/em/tsconfig.eslint.json'
> > ```
> > 
> > 
> > It runs fine on the `dev` branch, and this PR is up-to-date with `dev`. The error itself is nonsensical since `/tsconfig.eslint.json` does exist. Not to mention that it doesn't seem to have anything to do with `./src/reducers/editingValue.js`.
> > Would be so grateful for you to investigate!
> 
> Hi Raine! It's a very strange issue. This PR is up-to-date with dev and has got all the changes related to typescript configuration. It works fine for me without any error. I tried looking for any possible ts configuration error but there are none. Have you tried updating the `typescript ` package to the latest version `^3.7.5` ?

@shresthabijay Updated my `npm`, deleted `npm_modules` and `package-lock.json`, merged with `dev`, and it is working now!

raineorshine commented:
> > > @shresthabijay I'm getting a very strange error when I try to run `npm start`:
> > > ```
> > > ./src/reducers/editingValue.js
> > >   Line 0:  Parsing error: ENOENT: no such file or directory, stat '/users/raine/projects/em/tsconfig.eslint.json'
> > > ```
> > > 
> > > 
> > > It runs fine on the `dev` branch, and this PR is up-to-date with `dev`. The error itself is nonsensical since `/tsconfig.eslint.json` does exist. Not to mention that it doesn't seem to have anything to do with `./src/reducers/editingValue.js`.
> > > Would be so grateful for you to investigate!
> > 
> > 
> > Hi Raine! It's a very strange issue. This PR is up-to-date with dev and has got all the changes related to typescript configuration. It works fine for me without any error. I tried looking for any possible ts configuration error but there are none. Have you tried updating the `typescript ` package to the latest version `^3.7.5` ?
> 
> @shresthabijay Updated my `npm`, deleted `npm_modules` and `package-lock.json`, merged with `dev`, and it is working now!

And reappeared again after switching to another branch before 7b884090a6c0a6ef2a728fcdfb68b3d3fddd87a6.

Turns out the eslint cache can get out of sync, [as described here](https://github.com/webpack-contrib/eslint-loader/issues/214).

Fix with `rm -rf node_modules/.cache/eslint-loader`
